### PR TITLE
More transaction service tests ported to the new test tool

### PR DESCRIPTION
**Pull request by stefanobaghino-da**
This should be the second to last PR that contributes #2838 before the effort can be considered finished.

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [x] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Add a line to the [release notes](https://github.com/digital-asset/daml/blob/master/unreleased.rst), if appropriate
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/infrastructure/participant/ParticipantTestContext.scala [hunk 1]
```diff
@@ -286,11 +296,75 @@ private[testtool] final class ParticipantTestContext private[participant] (
...
+    */
+  def getTransactionByIdRequest(
+      transactionId: String,
+      parties: Seq[Party]): GetTransactionByIdRequest =
+    new GetTransactionByIdRequest(ledgerId, transactionId, Tag.unsubst(parties))
+
+  /**
+    * Non-managed version of [[transactionTreeById]], use this only if you need to tweak the request (i.e. to test low-level details)
+    */
+  def transactionTreeById(request: GetTransactionByIdRequest): Future[TransactionTree] =
```
Comment (gerolf-da) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/infrastructure/participant/ParticipantTestContext.scala]: Do you want to move the "low-level" testing methods into a nested object?
Something like:
```
class ParticipantTestContext {
  object lowLevel {
    // low-level methods here
  }
```
This way it is fairly obvious that these are discouraged from general purpose use. What do you think?

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/TransactionService.scala [hunk 2]
```diff
@@ -806,6 +807,523 @@ class TransactionService(session: LedgerSession) extends LedgerTestSuite(session
...
+      "Expose the same events for each transaction as the output of getTransactionTrees") {
+      context =>
+        for {
+          Vector(alpha, beta) <- context.participants(2)
+          submitter <- alpha.allocateParty()
+          listener <- beta.allocateParty()
+          _ <- alpha.create(submitter, AgreementFactory(listener, submitter))
+          _ <- synchronize(alpha, beta)
+          trees <- alpha.transactionTrees(listener, submitter)
+          _ <- eventually {
```
Comment (gerolf-da) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/TransactionService.scala]: I don't think the `eventually` is needed, considering we have a synchronization point before this.

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/TransactionService.scala [hunk 3]
```diff
@@ -806,6 +807,523 @@ class TransactionService(session: LedgerSession) extends LedgerTestSuite(session
...
+        }
+    }
+
+  private[this] val flatTransactionById =
+    LedgerTest("TXFlatTransactionById", "Expose a visible transaction by identifier") { context =>
+      for {
+        ledger <- context.participant()
+        party <- ledger.allocateParty()
+        dummy <- ledger.create(party, Dummy(party))
+        tree <- ledger.exerciseForFlatTransaction(party, dummy.exerciseDummyChoice1)
```
Comment (gerolf-da) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/TransactionService.scala]: `exerceForFlatTransaction` returns a ... flat transaction and not a tree?
```suggestion
        flat <- ledger.exerciseForFlatTransaction(party, dummy.exerciseDummyChoice1)
```

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/TransactionService.scala [hunk 4]
```diff
@@ -806,6 +807,523 @@ class TransactionService(session: LedgerSession) extends LedgerTestSuite(session
...
+        byId <- ledger.flatTransactionById(tree.transactionId, party)
```
Comment (gerolf-da) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/TransactionService.scala]:
```suggestion
        byId <- ledger.flatTransactionById(flat.transactionId, party)
```

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/TransactionService.scala [hunk 5]
```diff
@@ -806,6 +807,523 @@ class TransactionService(session: LedgerSession) extends LedgerTestSuite(session
...
+      } yield {
+        assertEquals("The transaction fetched by identifier does not match", tree, byId)
```
Comment (gerolf-da) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/TransactionService.scala]:
```suggestion
        assertEquals("The transaction fetched by identifier does not match", flat, byId)
```

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/TransactionService.scala [hunk 6]
```diff
@@ -806,6 +807,523 @@ class TransactionService(session: LedgerSession) extends LedgerTestSuite(session
...
+    LedgerTest(
+      "TXFlatTransactionByIdSameAsTransactionStream",
+      "Expose the same events for each transaction as the output of getTransactions") { context =>
+      for {
+        Vector(alpha, beta) <- context.participants(2)
+        submitter <- alpha.allocateParty()
+        listener <- beta.allocateParty()
+        _ <- alpha.create(submitter, AgreementFactory(listener, submitter))
+        _ <- synchronize(alpha, beta)
+        trees <- alpha.flatTransactions(listener, submitter)
```
Comment (gerolf-da) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/TransactionService.scala]:
```suggestion
        flats <- alpha.flatTransactions(listener, submitter)
```

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/TransactionService.scala [hunk 7]
```diff
@@ -806,6 +807,523 @@ class TransactionService(session: LedgerSession) extends LedgerTestSuite(session
...
+        _ <- eventually {
```
Comment (gerolf-da) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/TransactionService.scala]: Similar to the comment further above, I don't think we need the `eventually` block here, do we?

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/TransactionService.scala [hunk 8]
```diff
@@ -806,6 +807,523 @@ class TransactionService(session: LedgerSession) extends LedgerTestSuite(session
...
+    }
+
+  private[this] val transactionTreeByEventIdInvalid =
+    LedgerTest(
+      "TXTransactionTreeByEventIdInvalid",
+      "Return INVALID when looking up an invalid transaction tree by event identifier") { context =>
+      for {
+        ledger <- context.participant()
+        party <- ledger.allocateParty()
+        failure <- ledger.transactionTreeByEventId("dont' worry, be happy", party).failed
```
Comment (gerolf-da) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/TransactionService.scala]: 😄

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/TransactionService.scala [hunk 9]
```diff
@@ -806,6 +807,523 @@ class TransactionService(session: LedgerSession) extends LedgerTestSuite(session
...
+
+  private[this] val flatTransactionByEventId =
+    LedgerTest(
+      "TXFlatTransactionByEventId",
+      "Expose a visible flat transaction by event identifier") { context =>
+      for {
+        ledger <- context.participant()
+        party <- ledger.allocateParty()
+        dummy <- ledger.create(party, Dummy(party))
+        tree <- ledger.exerciseForFlatTransaction(party, dummy.exerciseDummyChoice1)
```
Comment (gerolf-da) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/TransactionService.scala]:
```suggestion
        flat <- ledger.exerciseForFlatTransaction(party, dummy.exerciseDummyChoice1)
```

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/TransactionService.scala [hunk 10]
```diff
@@ -806,6 +807,523 @@ class TransactionService(session: LedgerSession) extends LedgerTestSuite(session
...
+        event = tree.events.head.event
+        eventId = event.created.map(_.eventId).getOrElse(event.archived.map(_.eventId).get)
```
Comment (gerolf-da) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/TransactionService.scala]: Hm, shouldn't we know precisely which event we expect? Although I guess in this case it doesn't matter so much, as we simply want to look up the transaction by any eventId.

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/TransactionService.scala (hunk 8 above)
Comment (stefanobaghino-da) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/TransactionService.scala]: Literally coming from the existing implementation. :wink:

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/infrastructure/participant/ParticipantTestContext.scala (hunk 1 above)
Comment (stefanobaghino-da) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/infrastructure/participant/ParticipantTestContext.scala]: I was thinking of the same and I'm in favor of this (also to keep the "actual" API a bit leaner and cleaner). I would address this as part of a different ticket though. Is that OK with you?

Comment (gerolf-da) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/infrastructure/participant/ParticipantTestContext.scala]: Sure, thanks 👍

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/TransactionService.scala (hunk 10 above)
Comment (stefanobaghino-da) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/TransactionService.scala]: I believe the intent here was to not encode the order in which a transaction arrived in the test. I'm not sure this makes sense though and I would be in favor of pinning the expected transaction to be either a create or an archival. Makes sense?

Comment (gerolf-da) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/TransactionService.scala]: Right. The Ledger API doesn't specify the order or the events, but there should only be a single archived event in this case anyway.

Comment (stefanobaghino-da) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/TransactionService.scala]: Of course, thanks.

Comment (stefanobaghino-da) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/TransactionService.scala]: Addressed by https://github.com/digital-asset/daml/pull/2961/commits/962544588c5b708f7b2bd392347a46652d088d34

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/TransactionService.scala (hunk 9 above)
Comment (stefanobaghino-da) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/TransactionService.scala]: Addresse by https://github.com/digital-asset/daml/pull/2961/commits/6b70633a47b0894876eadf4f62359fd2b5560c31

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/TransactionService.scala (hunk 6 above)
Comment (stefanobaghino-da) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/TransactionService.scala]: Addressed by https://github.com/digital-asset/daml/pull/2961/commits/6750f3224d1b6f137e929878d46bee4a1f835aab

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/TransactionService.scala (hunk 3 above)
Comment (stefanobaghino-da) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/TransactionService.scala]: Addressed by https://github.com/digital-asset/daml/pull/2961/commits/bd5b1d34a81387a3a4e686ea73806f73fd1e1ea1

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/TransactionService.scala (hunk 5 above)
Comment (stefanobaghino-da) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/TransactionService.scala]: Addressed by https://github.com/digital-asset/daml/pull/2961/commits/bd5b1d34a81387a3a4e686ea73806f73fd1e1ea1

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/TransactionService.scala (hunk 4 above)
Comment (stefanobaghino-da) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/TransactionService.scala]: Addressed by https://github.com/digital-asset/daml/pull/2961/commits/bd5b1d34a81387a3a4e686ea73806f73fd1e1ea1

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/TransactionService.scala (hunk 7 above)
Comment (stefanobaghino-da) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/TransactionService.scala]: You are right, probably a leftover from an earlier approach, thanks.

Comment (stefanobaghino-da) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/TransactionService.scala]: Addressed by https://github.com/digital-asset/daml/pull/2961/commits/b81eac6b9267070c0f30cb2f09e7fc8444857300

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/TransactionService.scala (hunk 2 above)
Comment (stefanobaghino-da) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/TransactionService.scala]: Addressed by https://github.com/digital-asset/daml/pull/2961/commits/b32f1221504aed4a2bc42e4068e5ee9f8fed8c49

**Pull request merged**